### PR TITLE
[EventEngine] Update Endpoint API to return true if read/write succeeds immediately

### PR DESCRIPTION
**Pull request by Vignesh2208**
Requires cherrypick

### CODE REVIEW

**Review by drfloob — COMMENTED** [review]
Looks good, thanks for doing this Vignesh!

File: src/core/lib/event_engine/posix_engine/posix_endpoint.cc [hunk 1]
```diff
@@ -1114,11 +1137,15 @@ void PosixEndpointImpl::Write(
...
     TcpShutdownTracedBufferList();
-    return;
+    if (handle_->IsHandleShutdown()) {
+      status = TcpAnnotateError(absl::InternalError("EOF"));
+      engine_->Run([on_writable = std::move(on_writable), status]() mutable {
+        on_writable(status);
+      });
+      return false;
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/posix_engine/posix_endpoint.cc:1148]: In the iomgr code, a write with 0 data isn't necessarily considered an error unless the endpoint is shutdown. Not sure what could happen if we break that semantics

File: src/core/lib/iomgr/event_engine_shims/endpoint.cc [hunk 2]
```diff
@@ -93,15 +94,98 @@ class EventEngineEndpointWrapper {
...
+  bool Read(grpc_closure* read_cb, grpc_slice_buffer* pending_read_buffer,
             const EventEngine::Endpoint::ReadArgs* args) {
-    endpoint_->Read(std::move(on_read), buffer, args);
+    Ref();
+    pending_read_cb_ = read_cb;
+    pending_read_buffer_ = pending_read_buffer;
+    // TODO(vigneshbabu): Use SliceBufferCast<> here.
+    SliceBuffer* read_buffer = new (&eeep_->read_buffer)
+        SliceBuffer(SliceBuffer::TakeCSliceBuffer(*pending_read_buffer_));
+    read_buffer->Clear();
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/iomgr/event_engine_shims/endpoint.cc:105]: The endpoint read implementations can only append to the read buffer. So we have to clear the read buffer to ensure the callback gets the correct contents. This is done by the iomgr endpoint code as well.

File: src/core/lib/event_engine/posix_engine/posix_endpoint.cc [hunk 3]
```diff
@@ -1142,13 +1169,20 @@ void PosixEndpointImpl::Write(
     write_cb_ = std::move(on_writable);
     current_zerocopy_send_ = zerocopy_send_record;
     handle_->NotifyOnWrite(on_write_);
+    return false;
   } else {
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/posix_engine/posix_endpoint.cc]: Fixed.

File: src/core/lib/event_engine/posix_engine/posix_endpoint.cc [hunk 4]
```diff
@@ -1142,13 +1169,20 @@ void PosixEndpointImpl::Write(
...
-      on_writable(status);
-    });
+    if (!status.ok()) {
+      // Write failed immediately. Schedule the on_writable callback to run
+      // asynchronously.
+      engine_->Run([on_writable = std::move(on_writable), status]() mutable {
+        on_writable(status);
+      });
+      return false;
+    } else {
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/posix_engine/posix_endpoint.cc]: Fixed.

File: src/core/lib/iomgr/event_engine_shims/endpoint.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/iomgr/event_engine_shims/endpoint.cc:105]: > The endpoint read implementations can only append to the read buffer.

Interesting. We didn't document it that way (in iomgr either), and there are a few spots that make other assumptions about the state of buffers before and after calling read. https://github.com/grpc/grpc/blob/dd07fd8669d4dc0a8c51f4d8841e1d6c8710b371/src/core/lib/transport/http_connect_handshaker.cc#L249-L255

We could stand to save a few allocations by reusing any slices in the buffer as "sacrificial", but we can experiment later.

File: src/core/lib/event_engine/posix_engine/posix_endpoint.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/event_engine/posix_engine/posix_endpoint.cc:1149]: That's fair. It seems like poor usage, but not egregiously so.

**Review by drfloob — APPROVED** [review]

File: include/grpc/event_engine/event_engine.h [hunk 5]
```diff
@@ -162,8 +162,11 @@ class EventEngine : public std::enable_shared_from_this<EventEngine> {
     /// Reads data from the Endpoint.
     ///
     /// When data is available on the connection, that data is moved into the
-    /// \a buffer, and the \a on_read callback is called. The caller must ensure
-    /// that the callback has access to the buffer when executed later.
+    /// \a buffer, and the \a on_read callback is called. If the read succeeds
```
Comment (ctiller) [inline include/grpc/event_engine/event_engine.h:165]: "and the \a on_read callback is called" should probably be removed

File: src/core/lib/iomgr/event_engine_shims/endpoint.cc [hunk 6]
```diff
@@ -93,15 +94,98 @@ class EventEngineEndpointWrapper {
...
   // Read using the underlying EventEngine endpoint object.
-  void Read(absl::AnyInvocable<void(absl::Status)> on_read, SliceBuffer* buffer,
+  bool Read(grpc_closure* read_cb, grpc_slice_buffer* pending_read_buffer,
             const EventEngine::Endpoint::ReadArgs* args) {
-    endpoint_->Read(std::move(on_read), buffer, args);
+    Ref();
+    pending_read_cb_ = read_cb;
+    pending_read_buffer_ = pending_read_buffer;
+    // TODO(vigneshbabu): Use SliceBufferCast<> here.
+    SliceBuffer* read_buffer = new (&eeep_->read_buffer)
```
Comment (ctiller) [inline src/core/lib/iomgr/event_engine_shims/endpoint.cc:103]: nit: prefer `Construct(&eeep_->read_buffer, SliceBuffer::TakeCSliceBuffer(...))`

**Review by ctiller — APPROVED** [review]

File: src/core/lib/iomgr/event_engine_shims/endpoint.cc (hunk 6 above)
Comment (Vignesh2208) [reply src/core/lib/iomgr/event_engine_shims/endpoint.cc]: Fixed.

File: include/grpc/event_engine/event_engine.h (hunk 5 above)
Comment (Vignesh2208) [reply include/grpc/event_engine/event_engine.h]: Removed.

**Pull request merged**